buildpress/src/main/scala/buildpress/io/SbtProjectHasher.scala

```diff
@@ -0,0 +1,78 @@
+package buildpress.io
+
+import java.io.IOException
+import java.nio.file._
+import java.nio.file.attribute.BasicFileAttributes
+import bloop.io.{AbsolutePath, ByteHasher}
+import buildpress.config.Config.{BuildSettingsHashes, HashedPath}
+
+object SbtProjectHasher {
+  class SbtFileMatcher(root: Path) extends PathMatcher {
+    override def matches(path: Path): Boolean = {
+      val projectDir: Path = root.resolve("project")
+      val projectMetaDir: Path = projectDir.resolve("project")
+      val parent: Path = path.getParent
+
+      (parent == root || parent == projectDir || parent == projectMetaDir) && {
+        val filename: String = path.getFileName.toString
+        filename == "build.properties" ||
+        filename.endsWith(".sbt") ||
+        filename.endsWith(".scala")
+      }
+    }
+  }
+
+  def hashProjectSettings(root: AbsolutePath): BuildSettingsHashes = {
+    val individual: List[(AbsolutePath, Int)] =
+      SbtProjectHasher.findAndHashSbtFilesInRepo(root.underlying)
+    BuildSettingsHashes(individual.map(ph => HashedPath(ph._1.underlying, ph._2)))
+  }
+
+  // TODO: hash structure, not plain files
+  def findAndHashSbtFilesInRepo(
+      path: Path
+  ): List[(AbsolutePath, Int)] = {
+    val collected = List.newBuilder[Path]
+    val pm = new SbtFileMatcher(path)
+    val discovery = new FileVisitor[Path] {
+      def visitFile(file: Path, attributes: BasicFileAttributes): FileVisitResult = {
+        if (pm.matches(file)) {
+          collected += file
+        }
+        FileVisitResult.CONTINUE
+      }
+
+      def visitFileFailed(
+          t: Path,
+          e: IOException
+      ): FileVisitResult = {
+        FileVisitResult.CONTINUE
+      }
+
+      def preVisitDirectory(
+          directory: Path,
+          attributes: BasicFileAttributes
+      ): FileVisitResult = {
+        FileVisitResult.CONTINUE
+      }
+
+      def postVisitDirectory(
+          directory: Path,
+          exception: IOException
+      ): FileVisitResult = {
+        FileVisitResult.CONTINUE
+      }
+    }
+
+    Files.walkFileTree(
+      path,
+      java.util.EnumSet.of(FileVisitOption.FOLLOW_LINKS),
+      Int.MaxValue,
```

Comment:
Shouldn't the depth level be here 1?

Reply:
practically yes, i haven't seen people going more than 1 `project/` level nesting, maybe I'll put 2 just to be safe, but this is a good catch performance-wise!

Reply:
I'm not sure sbt can read sbt files deeper than 1 level, especially in these meta builds, so I'd probably default on 1 :smile: